Remove ExceptionHelper.detailedMessage

pugnascotia opened the pull request:
Closes #19069.

I'm unsure of my approach here, so more opinions would be welcome. Mostly I replaced calls to `ExceptionHelper.detailedMessage` with `e.getMessage()`, which is all well and good but does remove some level of detail. There were a couple of cases where I replaced the method call with a `throw`, in the hope that code elsewhere will catch it and do the right thing.

pugnascotia committed:
Remove ExceptionHelper.detailedMessage

Closes #19069.

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst commented:
I don't think changing these uses to `getMessage()` solves the problem outlined in #19069. We often wrap exceptions, creating deep chains. By changing to getMessage() we omit even more information than detailedMessage is emitting. The problem is we need the actual stack trace. detailedMessage tries to summarize an exception into a single string that omits stack frames from each cause, summarizing as just the exception class + message for each cause. For something like a NullPointerException we lose all information that makes the exception debuggable (the location where the NPE occurred). I think a better substitute would be `ExceptionHelper.stackTrace`?

pugnascotia committed:
Use ExceptionsHelper.stackTrace instead of Exception.getMessage

pugnascotia commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/FailedShard.java:
```diff
@@ -42,8 +42,8 @@ public FailedShard(ShardRouting routingEntry, String message, Exception failure,
-        return "failed shard, shard " + routingEntry + ", message [" + message + "], failure [" +
-                   ExceptionsHelper.detailedMessage(failure) + "], markAsStale [" + markAsStale + "]";
+        return "failed shard, shard " + routingEntry + ", message [" + message + "], markAsStale [" + markAsStale + "], failure ["
+            + ExceptionsHelper.stackTrace(failure) + "]";
```
I changed the order of fields here so that the stack trace comes last.

pugnascotia commented on server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:
```diff
@@ -444,10 +444,10 @@ public String toString() {
+            components.add("markAsStale [" + markAsStale + "]");
             if (failure != null) {
-                components.add("failure [" + ExceptionsHelper.detailedMessage(failure) + "]");
+                components.add("failure [" + ExceptionsHelper.stackTrace(failure) + "]");
```
Also changed the order of elements here, so the stack trace is last.

rjernst commented on server/src/main/java/org/elasticsearch/ElasticsearchException.java:
```diff
@@ -681,7 +681,7 @@ public String toString() {
             builder.append(' ');
         }
-        return builder.append(ExceptionsHelper.detailedMessage(this).trim()).toString();
+        return builder.append(super.toString().trim()).toString();
```
Won't we lose all stack information by just using the exception tostring?

pugnascotia replied:
Only if whatever's calling `ElasticsearchException.toString()` isn't also printing the stack trace. The problem with calling `ElasticsearchException.stackTrace()` here is that it calls `.toString()` on the exception, so you end up blowing the stack. `detailedMessage()` didn't do this because it essentially reimplemented `Throwable.toString()`. Open to ideas here.

rjernst commented on server/src/main/java/org/elasticsearch/search/suggest/phrase/DirectCandidateGeneratorBuilder.java:
```diff
@@ -490,7 +490,7 @@ public String toString() {
             return Strings.toString(builder);
         } catch (Exception e) {
-            return "{ \"error\" : \"" + ExceptionsHelper.detailedMessage(e) + "\"}";
+            throw new RuntimeException(e);
```
Do we have a test for this failure case so we know the response is not already too far written to throw/serialize an exception like this?

pugnascotia replied:
We do have integration tests in `SearchSuggestIT` with failure cases, but they don't look related to a failure to serialize content. I'd be happy to add a new test - any ideas how to provoke a failure when serialising a `DirectCandidateGeneratorBuilder` to XContent?

rjernst reviewed (commented):
Thanks, this looks better to me. I left a couple last questions, and would also appreicate @jasontedor's opinion on whether using a full stack trace is always appropriate (IMO compared to the useless/reduced info from detailedmessage stacktrace is better).

rjernst reviewed (approved):
LGTM once CI passes

pugnascotia committed:
Fix checkstyle error

pugnascotia committed:
Unit test fixes

pugnascotia committed:
Yet more fixes

pugnascotia committed:
More test fixes

pugnascotia committed:
Another raft of fixes

pugnascotia commented on server/src/test/java/org/elasticsearch/action/admin/indices/close/CloseIndexResponseTests.java:
```diff
@@ -113,7 +113,9 @@ protected void assertEqualInstances(CloseIndexResponse expected, CloseIndexRespo
-                            assertThat(actualFailure.reason(), equalTo(expectedFailure.reason()));
+                            // Serialising and deserialising an exception seems to remove the "java.base/" part from the stack trace,
+                            // so these string replacements account for this.
+                            assertThat(actualFailure.reason().replace("java.base/", ""), equalTo(expectedFailure.reason().replace("java.base/", "")));
```
I made this change (and again in 1 other class) to get the test to pass, but I don't like it. I'd appreciate other suggestions.

pugnascotia committed:
Checkstyle again

pugnascotia committed:
Checkstyle

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 19069-remove-exceptionhelper-detailedmessage

rjernst commented on server/src/test/java/org/elasticsearch/action/admin/indices/close/CloseIndexResponseTests.java:
```diff
@@ -113,7 +113,11 @@ protected void assertEqualInstances(CloseIndexResponse expected, CloseIndexRespo
+                            // Serialising and deserialising an exception seems to remove the "java.base/" part from the stack trace,
+                            // so these string replacements account for this.
+                            assertThat(
+                                actualFailure.reason().replace("java.base/", ""),
```
Can we check for a portion of the failure instead of an exact stack trace?

pugnascotia replied:
Yes, I've reworked both instances of this.

pugnascotia committed:
Lots of test fixes

pugnascotia commented on server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java:
```diff
@@ -368,7 +368,7 @@ private void internalRecoverFromStore(IndexShard indexShard) throws IndexShardRe
                         files = Arrays.toString(store.directory().listAll());
                     } catch (Exception inner) {
                         inner.addSuppressed(e);
```
Agreed, I've removed this line.

pugnascotia committed:
Checkstyle

pugnascotia committed:
Rework messages in RestoreInProgressAllocationDecider

A message in RestoreInProgressAllocationDecider included details about
allocation decisions. If an allocation couldn't happen due to an
exception, this resulted in a stack trace in the middle of the message.
Move the details to the end of the message, which makes it easier to
handle in tests.

pugnascotia committed:
Checkstyle

pugnascotia committed:
Rewrite DetailedErrorsEnabledIT to parse responses

Instead of relying on substring matching, parse the response JSON and
check fields explicitly. This uses Jackson, which has been added to the
build dependencies for :qa:smoke-test-http.

pugnascotia committed:
Integration test fixes

This required adding a getting to `VerificationFailure` so that the
`cause` field could be accessed and added as suppressed exceptions
elsewhere.

pugnascotia committed:
Checkstyle

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 19069-remove-exceptionhelper-detailedmessage

pugnascotia commented:
@elasticsearch run elasticsearch-ci/1

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 19069-remove-exceptionhelper-detailedmessage

pugnascotia commented:
@rjernst I've merged in `master` again, would you mind taking another look?

rjernst commented on qa/smoke-test-http/build.gradle:
```diff
@@ -23,6 +23,7 @@ apply plugin: 'elasticsearch.rest-test'
 apply plugin: 'elasticsearch.test-with-dependencies'
 
 dependencies {
+    testCompile("com.fasterxml.jackson.core:jackson-databind:2.8.11")
```
nit: we normally do this without the parens so it maintains the "dsl" look

rjernst commented on server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java:
```diff
@@ -626,9 +631,23 @@ public void testMultiMatchQuery() throws Exception {
+            client().prepareSearch().setQuery(builder).get();
+            shardFailures = searchResponse.getShardFailures();
+            assertThat("Expected shard failures, got none", shardFailures, not(emptyArray()));
+        } catch (SearchPhaseExecutionException e) {
```
we should use expectThrows here

rjernst reviewed (approved):
LGTM, two last nits

pugnascotia committed:
Fixes from review

pugnascotia commented:
@elasticmachine run elasticsearch-ci/1

pugnascotia commented:
@elasticmachine run elasticsearch-ci/bwc

pugnascotia commented:
@elasticmachine run elasticsearch-ci/bwc

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 19069-remove-exceptionhelper-detailedmessage

pugnascotia commented:
@elasticmachine run elasticsearch-ci/2
